### PR TITLE
2018-04-26 wayland-rs 0.20

### PR DESCRIPTION
**Pull request by elinorbgr**

### CODE REVIEW

File: content/wayland-rs-v0.20.markdown [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+small runtime costs for large ergonomic improvements". The previous design of the libraries
+tried to remain as low-overhead possible as possible on the wayland C libs. This caused an
+API that was quite convoluted to try to
+[make safe sharing of mutable state](https://docs.rs/token_store/), and at 
+some points not very different from manually re-implementing a v-table. Other issues more
+specific to smithay resulted in a large accumulation of template parameters on types that need
+to be named and stored. All these issues made the API quite tedious to use overall.
+
+As a result, for this new version I decided to embrace trait objects, `Rc`, and `Arc`.
+Additionnaly, the API redesign was made focusing on *the spirit* of the wayland protocol,
```
Comment (Drakulix) [inline content/wayland-rs-v0.20.markdown]: Additionnaly

File: content/wayland-rs-v0.20.markdown [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+  solved (see next section)
+- The ergonomics and flexibility of specifying callbacks for wayland events has been greatly
+  improved
+- Overall better interaction with objects from the C world, if wayland-rs is used in
+  combination with C libraries
+
+## The new callback specification
+
+The most major API change is on the way callbacks are handled. Previously you had to provide
+a set of freestanding functions that would be given acces to some state data for each event,
```
Comment (Drakulix) [inline content/wayland-rs-v0.20.markdown]: acces

File: content/wayland-rs-v0.20.markdown [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+I also took the occasion of this large API rework to fuse
+[wayland-kbd](https://github.com/smithay/wayland-kbd) and
+[wayland-window](https://github.com/smithay/wayland-window) into a single crate named
+[smithay-client-toolkit](https://github.com/smithay/client-toolkit), which will be some
+kind of an equivalent of [smithay](https://github.com/smithay/smithay), but for client
+applications. Given the scope of what is possible, it'll clearly remain much simpler than
+smithay itself, as it only abstracts the wayland protocol, and nothing like all of
+smithay's backends.
+
+Currently, it only provides some barebones functionnalities, but there is still a lot
```
Comment (Drakulix) [inline content/wayland-rs-v0.20.markdown]: functionnalities

**Review by Drakulix — COMMENTED** [review]
Found some typos.

You might want to include all the work that went into potentially allowing wayland-rs to run without C-libraries in the future, no?

File: content/wayland-rs-v0.20.markdown [hunk 4]
```diff
@@ -27,7 +27,7 @@ specific to smithay resulted in a large accumulation of template parameters on t
 to be named and stored. All these issues made the API quite tedious to use overall.
 
 As a result, for this new version I decided to embrace trait objects, `Rc`, and `Arc`.
-Additionnaly, the API redesign was made focusing on *the spirit* of the wayland protocol,
+Additionaly, the API redesign was made focusing on *the spirit* of the wayland protocol,
```
Comment (jplatte) [inline content/wayland-rs-v0.20.markdown]: It's additionally, with two `l`

File: content/wayland-rs-v0.20.markdown [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+## The new callback specification
+
+The most major API change is on the way callbacks are handled. Previously you had to provide
+a set of freestanding functions that would be given access to some state data for each event,
+with the difficulties of sharing state between an undeterminate set of callbacks a priori...
+
+Now each time a new protocol object is created, you'll receive it as a `NewProxy`/`NewResource`
+object (depending on whether you are client side or server side), and will have to implement
+it before accessing the real `Proxy`/`Resource` and being able to use it. Implementing an
+object consists in providing a type implementing the appropriate `Implementation<Meta, Msg>`
```
Comment (jplatte) [inline content/wayland-rs-v0.20.markdown]: consists *of*

File: content/wayland-rs-v0.20.markdown [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+```
+
+Here the `Meta` type parameter represent some metadata associated with the message you
+receive (often a `Proxy`/`Resource` handle to the wayland object receiving the message),
+while `Msg` is a type representing the message itself (often an enum of the possible wayland
+messages this object can receive).
+
+The convenience is that `Implementation<Meta, Msg>` is automatically implemented for all
+type that implement `FnMut(Msg, Meta)`. As such, you can easily provide closures as
+implementations, with all the ergonomics of capturing values and code compacity that
```
Comment (jplatte) [inline content/wayland-rs-v0.20.markdown]: "compacity" doesn't exist. There is "compactness", but "terseness" would probably be better.

File: content/wayland-rs-v0.20.markdown [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+implies. The only catch being that the implementations must be `'static`, so you'll need
+to rely on `Arc` or `Rc` to share data. The runtime cost implied should be negligible however,
+the wayland protocol part should hardly be a bottleneck for any application.
+
+There are also two ways to implement an object, the default one requiring from the 
+implementation to be `Send`, and a secondary one that relaxes this requirement provided you
```
Comment (jplatte) [inline content/wayland-rs-v0.20.markdown]: requiring ~~from~~ the implementation *type* to be `Send`

(adding *type* is of course not something grammatical, but given that "implementation" usually refers to a function, I think it makes sense for the type impl'ing `Implementation` [that trait has a really weird name...] to be referred to as "the implementation type")

File: content/wayland-rs-v0.20.markdown [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+use sctk::reexports::client::{Display, Proxy};
+use sctk::reexports::client::protocol::wl_display::RequestsTrait as DisplayRequests;
+use sctk::reexports::client::protocol::wl_compositor::RequestsTrait as CompositorRequests;
+
+fn main() {
+    // Connect to a wayland server
+    let (display, mut event_queue) = Display::connect_to_env().unwrap();
+
+    // The Environment is an abstraction binding for you most of the
+    // classic globals. Otherwise this job is very boring and repetitive.
```
Comment (jplatte) [inline content/wayland-rs-v0.20.markdown]: [...] an abstraction that binds most of the classic globals for you. Doing this manually is very boring and repetitive.

File: content/wayland-rs-v0.20.markdown [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+    // contents of our window
+    let surface = env.compositor
+        .create_surface()
+        .unwrap()
+        .implement(|_, _| {
+            /* this is the surface implementation, this one ignores all events */
+        });
+    
+    // Now create a Window for this surface
+    // The window abstracts for all the protocol handling of the shell, and provides
```
Comment (jplatte) [inline content/wayland-rs-v0.20.markdown]: The window abstracts ~~for~~ all the protocol handling of the shell

(probably?)

File: content/wayland-rs-v0.20.markdown [hunk 10]
```diff
@@ -41,6 +41,21 @@ To list a few clear gains from this rework:
 - Overall better interaction with objects from the C world, if wayland-rs is used in
   combination with C libraries
 
+An other interesting point is that, as this rework is much less centered around the C API,
```
Comment (jplatte) [inline content/wayland-rs-v0.20.markdown]: Another

File: content/wayland-rs-v0.20.markdown [hunk 11]
```diff
@@ -41,6 +41,21 @@ To list a few clear gains from this rework:
...
+it allowed me to keep open the door for a full-rust implementation of the protocol. Thus I
+conceived the libraries around a cargo feature `native_lib`, controlling whether the native
```
Comment (jplatte) [inline content/wayland-rs-v0.20.markdown]: Not sure what you meant here, but "conceive" doesn't really make sense in this context.

**Pull request merged**